rustc suggests using enum type as variant

Both the error and the suggestion make it seem like `Enum` is an enum variant; it's a type, not a variant.

```rust
pub enum Enum {
    V1(i32),
}

pub fn foo(x: i32) -> Enum {
    Enum::V1 { x }
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=981550655b719d36af302db57a6acc9f))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0559]: variant `Enum::V1` has no field named `x`
 --> src/lib.rs:6:16
  |
2 |     V1(i32),
  |     -- `Enum` defined here
...
6 |     Enum::V1 { x }
  |     --------   ^ field does not exist
  |     |
  |     `Enum` is a tuple variant, use the appropriate syntax: `Enum(/* fields */)`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0559`.
error: could not compile `playground`

To learn more, run the command again with --verbose.

```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"bugadani"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_typeck/src/check/expr.rs
+++ b/compiler/rustc_typeck/src/check/expr.rs
@@ -1381,19 +1381,42 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             ty,
         );
         match variant.ctor_kind {
-            CtorKind::Fn => {
-                err.span_label(variant.ident.span, format!("`{adt}` defined here", adt = ty));
-                err.span_label(field.ident.span, "field does not exist");
-                err.span_label(
-                    ty_span,
-                    format!(
-                        "`{adt}` is a tuple {kind_name}, \
-                         use the appropriate syntax: `{adt}(/* fields */)`",
-                        adt = ty,
-                        kind_name = kind_name
-                    ),
-                );
-            }
+            CtorKind::Fn => match ty.kind() {
+                ty::Adt(adt, ..) if adt.is_enum() => {
+                    err.span_label(
+                        variant.ident.span,
+                        format!(
+                            "`{adt}::{variant}` defined here",
+                            adt = ty,
+                            variant = variant.ident,
+                        ),
+                    );
+                    err.span_label(field.ident.span, "field does not exist");
+                    err.span_label(
+                        ty_span,
+                        format!(
+                            "`{adt}::{variant}` is a tuple {kind_name}, \
+                             use the appropriate syntax: `{adt}::{variant}(/* fields */)`",
+                            adt = ty,
+                            variant = variant.ident,
+                            kind_name = kind_name
+                        ),
+                    );
+                }
+                _ => {
+                    err.span_label(variant.ident.span, format!("`{adt}` defined here", adt = ty));
+                    err.span_label(field.ident.span, "field does not exist");
+                    err.span_label(
+                        ty_span,
+                        format!(
+                            "`{adt}` is a tuple {kind_name}, \
+                                 use the appropriate syntax: `{adt}(/* fields */)`",
+                            adt = ty,
+                            kind_name = kind_name
+                        ),
+                    );
+                }
+            },
             _ => {
                 // prevent all specified fields from being suggested
                 let skip_fields = skip_fields.iter().map(|ref x| x.ident.name);

--- a/src/test/ui/issues/issue-80607.rs
+++ b/src/test/ui/issues/issue-80607.rs
@@ -4,7 +4,7 @@ pub enum Enum {
 }
 
 pub fn foo(x: i32) -> Enum {
-    Enum::V1 { x } //~ ERROR field does not exist
+    Enum::V1 { x } //~ ERROR `Enum::V1` has no field named `x`
 }
 
 fn main() {}

--- a/src/test/ui/issues/issue-80607.rs
+++ b/src/test/ui/issues/issue-80607.rs
@@ -1,0 +1,10 @@
+// This tests makes sure the diagnostics print the offending enum variant, not just the type.
+pub enum Enum {
+    V1(i32),
+}
+
+pub fn foo(x: i32) -> Enum {
+    Enum::V1 { x } //~ ERROR field does not exist
+}
+
+fn main() {}

--- a/src/test/ui/issues/issue-80607.stderr
+++ b/src/test/ui/issues/issue-80607.stderr
@@ -2,12 +2,12 @@ error[E0559]: variant `Enum::V1` has no field named `x`
   --> $DIR/issue-80607.rs:7:16
    |
 LL |     V1(i32),
-   |     -- `Enum` defined here
+   |     -- `Enum::V1` defined here
 ...
 LL |     Enum::V1 { x }
    |     --------   ^ field does not exist
    |     |
-   |     `Enum` is a tuple variant, use the appropriate syntax: `Enum(/* fields */)`
+   |     `Enum::V1` is a tuple variant, use the appropriate syntax: `Enum::V1(/* fields */)`
 
 error: aborting due to previous error
 

--- a/src/test/ui/issues/issue-80607.stderr
+++ b/src/test/ui/issues/issue-80607.stderr
@@ -1,0 +1,14 @@
+error[E0559]: variant `Enum::V1` has no field named `x`
+  --> $DIR/issue-80607.rs:7:16
+   |
+LL |     V1(i32),
+   |     -- `Enum` defined here
+...
+LL |     Enum::V1 { x }
+   |     --------   ^ field does not exist
+   |     |
+   |     `Enum` is a tuple variant, use the appropriate syntax: `Enum(/* fields */)`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0559`.

